src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -217,35 +217,12 @@ public ValueTask<FlushResult> WriteResponseTrailers(int streamId, HttpResponseTr
             }
         }
 
-        private static int EncodeStatusCode(int statusCode, Span<byte> buffer)
-        {
-            switch (statusCode)
-            {
-                // Status codes which exist in the HTTP/2 StaticTable.
-                case 200:
-                case 204:
-                case 206:
-                case 304:
-                case 400:
-                case 404:
-                case 500:
-                    buffer[0] = (byte)(0x80 | H2StaticTable.StatusIndex[statusCode]);
-                    return 1;
-                default:
-                    // Send as Literal Header Field Without Indexing - Indexed Name
-                    buffer[0] = 0x08;
-
-                    ReadOnlySpan<byte> statusBytes = System.Net.Http.HPack.StatusCodes.ToStatusBytes(statusCode);
-                    buffer[1] = (byte)statusBytes.Length;
-                    statusBytes.CopyTo(buffer.Slice(2));
-
-                    return 2 + statusBytes.Length;
-            }
-        }
-
         private static bool EncodeHeaders(int statusCode, Http2HeadersEnumerator headersEnumerator, Span<byte> buffer, out int length)
         {
-            var statusCodeLength = EncodeStatusCode(statusCode, buffer);
+            HPackEncoder.EncodeIndexedStatusHeader(statusCode, buffer, out var statusCodeLength);
```

Comment:
You're not checking the return value that ensures there was space. Of course there should always be enough space, it's the first header and it's tiny and if you can't write it then the response should throw.

Either check the return value and throw or have EncodeIndexedStatusHeader `throw new HPackEncodingException(SR.net_http_hpack_encode_failure);` like the other EncodeHeaders method does.